scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5217 @@
+        #     # I'm not sure if it's safe, though. If it causes a bug someday,
+        #     # I guess it wasn't.
+        #     # Even if there is no bug because of the shape, it is incorrect for
+        #     # `contains` to return True when there are invalid (e.g. NaN)
+        #     # parameters.
+        #     return np.asarray(True)
+
+        a, b = self.get_numerical_endpoints(parameter_values)
+        left_inclusive, right_inclusive = self.inclusive
+
+        in_left = item >= a if left_inclusive else item > a
+        in_right = item <= b if right_inclusive else item < b
+        return in_left & in_right
+
+
+class _RealDomain(_SimpleDomain):
+    r""" Represents a simply-connected subset of the real line.
+
+    Completes the implementation of the `_SimpleDomain` class for simple
+    domains on the real line.
+
+    Methods
+    -------
+    define_parameters(*parameters)
+        (Inherited) Records any parameters used to define the endpoints of the
+        domain.
+    get_numerical_endpoints(parameter_values)
+        (Inherited) Gets the numerical values of the domain endpoints, which
+        may have been defined symbolically.
+    contains(item, parameter_values)
+        (Inherited) Determines whether the argument is contained within the
+        domain
+    __str__()
+        Returns a string representation of the domain, e.g. "[a, b)".
+    draw(size, rng, proportions, parameter_values)
+        Draws random values based on the domain. Proportions of values within
+        the domain, on the endpoints of the domain, outside the domain,
+        and having value NaN are specified by `proportions`.
+
+    """
+
+    def __str__(self):
+        a, b = self.endpoints
+        left_inclusive, right_inclusive = self.inclusive
+
+        left = "[" if left_inclusive else "("
+        a = self.symbols.get(a, f"{a}")
+        right = "]" if right_inclusive else ")"
+        b = self.symbols.get(b, f"{b}")
+
+        return f"{left}{a}, {b}{right}"
+
+    def draw(self, size=None, rng=None, proportions=None, parameter_values=None):
+        r""" Draw random values from the domain.
+
+        Parameters
+        ----------
+        size : tuple of ints
+            The shape of the array of valid values to be drawn.
+        rng : np.Generator
+            The Generator used for drawing random values.
+        proportions : tuple of numbers
+            A tuple of four non-negative numbers that indicate the expected
+            relative proportion of elements that:
+
+            - are strictly within the domain,
+            - are at one of the two endpoints,
+            - are strictly outside the domain, and
+            - are NaN,
+
+            respectively. Default is (1, 0, 0, 0). The number of elements in
+            each category is drawn from the multinomial distribution with
+            `np.prod(size)` as the number of trials and `proportions` as the
```

Comment:
what does `draw` do in case the domain is unbounded? if it is bounded, my expectation is that the numbers in the domain are chosen uniformly.

Reply:
If you try to draw a random number from an infinite domain, you get an `OverflowError: high - low range exceeds valid bound`. This does not happen in the tests because the "typical" range is used by default rather than the whole domain, and that is always finite.

However, the goal is to create a more challenging benchmarking set that *does* sample from the floating-point representable domain with a specified distribution (e.g. equal probability of any floating point-representable value). We will instantiate two random variable objects with the sampled parameters: one that uses NumPy as the backend, and another that uses [`mparray`](https://github.com/mdhaber/mparray/) (or another array-API  compatible arbitrary precision array backend) as the backend for use as a reference implementation. We can compare the two values to get a sense of the error throughout the domain, and document the results. As an example, see e.g. [Boost's "Accuracy" documentation for `ibeta`](https://live.boost.org/doc/libs/1_39_0/libs/math/doc/sf_and_dist/html/math_toolkit/special/sf_beta/ibeta_function.html#math_toolkit.special.sf_beta.ibeta_function.accuracy). Of course, users could also use the arbitrary precision backend for their own calculations if the provided accuracy is insufficient.